bigtable_automation/gcf/bt_trigger.go

```diff
@@ -6,311 +20,237 @@
+	if parts[0] != "gs:" || parts[1] != "" || len(parts) < 3 {
+		return "", "", errors.Errorf("Unexpected path: %s", path)
 	}
+	return parts[2], strings.Join(parts[3:], "/"), nil
+}
 
-	bucket := gcsClient.Bucket(bucketName)
-	rc, err := bucket.Object(fileName).NewReader(ctx)
+func isObjectExist(ctx context.Context, path string) (bool, error) {
+	bucket, object, err := parsePath(path)
 	if err != nil {
-		log.Printf("Unable to open file from bucket %q, file %q: %v\n", bucketName, fileName, err)
-		return nil, status.Errorf(
-			codes.Internal, "Unable to open file from bucket %q, file %q: %v", bucketName, fileName, err)
+		return false, err
 	}
-	defer rc.Close()
-	return ioutil.ReadAll(rc)
-}
-
-func writeToGCS(ctx context.Context, bucketName, fileName, data string) error {
 	gcsClient, err := storage.NewClient(ctx)
 	if err != nil {
-		log.Printf("Failed to create gcsClient: %v\n", err)
-		return status.Errorf(codes.Internal, "Failed to create gcsClient: %v", err)
+		return false, errors.Wrap(err, "Failed to create gcsClient")
 	}
+	_, err = gcsClient.Bucket(bucket).Object(object).Attrs(ctx)
+	if err == storage.ErrObjectNotExist {
+		return false, nil
+	}
+	return true, nil
+}
 
-	bucket := gcsClient.Bucket(bucketName)
-	w := bucket.Object(fileName).NewWriter(ctx)
-	if _, err := fmt.Fprint(w, data); err != nil {
-		w.Close()
-		log.Printf("Unable to open file for writing from bucket %q, file %q: %v\n", bucketName, fileName, err)
-		return status.Errorf(codes.Internal, "Unable to write to bucket %q, file %q: %v", bucketName, fileName, err)
+func writeToGCS(ctx context.Context, path, data string) error {
+	bucket, object, err := parsePath(path)
+	if err != nil {
+		return err
+	}
+	gcsClient, err := storage.NewClient(ctx)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create gcsClient")
 	}
-	return w.Close()
+	w := gcsClient.Bucket(bucket).Object(object).NewWriter(ctx)
+	defer w.Close()
+	_, err = fmt.Fprint(w, data)
+	return err
 }
 
-func launchDataflowJob(ctx context.Context, env *environment, btInstance, cacheType, tableID string) error {
+func launchDataflowJob(
+	ctx context.Context,
+	projectID string,
+	instance string,
+	tableID string,
+	dataPath string,
+	controlPath string,
+	dataflowTemplate string,
+) error {
 	dataflowService, err := dataflow.NewService(ctx)
 	if err != nil {
-		log.Printf("Unable to create dataflow service: %v\n", err)
-		return status.Errorf(codes.Internal, "Unable to create dataflow service: %v", err)
+		return errors.Wrap(err, "Unable to create dataflow service")
 	}
-	inFile := fmt.Sprintf("gs://%s/%s/%s", env.dataBucket, tableID, dataFilePattern)
-	outFile := fmt.Sprintf("gs://%s/%s/%s/%s", env.controlBucket, cacheType, tableID, completedFile)
+	dataFile := fmt.Sprintf("%s/%s/%s", dataPath, tableID, dataFilePattern)
+	launchedPath := fmt.Sprintf("%s/%s/%s", controlPath, tableID, launchedFile)
+	completedPath := fmt.Sprintf("%s/%s/%s", controlPath, tableID, completedFile)
```

Comment:
dataPath and controlPath come from user input now.  Is a stale "/" OK?  Or should we use join() ?